[multitop] Add support for per-IP DT customization, add example for pwrmgr

pamaury opened the pull request:
Some IPs require more information than what is provided by a DT. An example is the `pwrmgr` which needs to list wake up sources. This PR refactors some of the HW desc and adds an extension mechanism to dtgen. IPs can opt-in to generate extra information by creating a python plugin. Add an example with the `pwrmgr`.

**pwrmgr extension design**

The first requirement to expose the necessary information is that the list of wakeups for each IP needs to be exposed. This is done by creating a new enum in each header with such a list. For example:
```c
// In dt_pinmux.h
/**
 * List of wakeups.
 *
 * Wakeups are guaranteed to be numbered consecutively from 0.
 */
typedef enum dt_pinmux_wakeup {
  kDtPinmuxWakeupPinWkupReq = 0, /**< pin wake request */
  kDtPinmuxWakeupUsbWkupReq = 1, /**< usb wake request */
  kDtPinmuxWakeupCount = 2, /**< Number of wakeups */
} dt_pinmux_wakeup_t;
```

Once this is done, we want the `pwrmgr` to expose to key pieces of informations:
- The list of wakeups and what they are: to identify a wakeup, we need to provide two pieces of information:
   - the instance ID that triggered the wakeup
   - the wakeup signal of that instance (some instances can have more than one, e.g. the pinmux)
- The list of reset requests and where they come from: the instance ID of module.

The API to query this information looks like this:
```c
// In dt_pwrmgr.h
/**
 * Description of a wakeup source.
 *
 * A wakeup source is always identified by the instance ID of the module where it comes from.
 * Some instances can have several wakeup signals, e.g. the pinmux has two (`pin` and `usb`).
 * For such IPs, it is not sufficient to know the instance, we also need to know which
 * signal triggered the wakeup. The `wakeup` index can be used to distinguish between those.
 * This value should be casted to the `dt_<ip>_wakeup_t` type of the corresponding IP.
 * For example, if the `pwrmgr` has two `pinmux` wakeup sources as described above, it's
 * two wakeup sources will be described as follows:
 * ```c
 * {.inst_id = kDtInstanceIdPinmux, .wakeup = kDtPinmuxWakeupPinWkupReq}, // for `pin`
 * {.inst_id = kDtInstanceIdPinmux, .wakeup = kDtPinmuxWakeupUsbWkupReq}, // for `usb`
 * ```
 */
typedef struct dt_pwrmgr_wakeup_src {
  dt_instance_id_t inst_id; /**< Instance ID of the source of this wakeup. */
  size_t wakeup; /**< Index of the wakeup signal for that instance. */
} dt_pwrmgr_wakeup_src_t;


/**
 * Get the number of wakeup sources.
 *
 * @param dt Instance of pwrmgr.
 * @return Number of wakeup sources.
 */
size_t dt_pwrmgr_wakeup_src_count(dt_pwrmgr_t dt);

/**
 * Get the description of a wakeup source.
 *
 * The wakeup sources are ordered as they appear in the registers.
 *
 * @param dt Instance of pwrmgr.
 * @param idx Index of the wakeup source, between 0 and `dt_pwrmgr_wakeup_src_count(dt)-1`.
 * @return Description of the source.
 */
dt_pwrmgr_wakeup_src_t dt_pwrmgr_wakeup_src(dt_pwrmgr_t dt, size_t idx);

/**
 * Description of a reset request source.
 *
 * A reset request source is always identified by the instance ID of the module where it comes from.
 * In principle, some instances could have several reset requests. If this is the case, the `rst_req`
 * can be used to distinguish between those. It should be cast to the `dt_<ip>_reset_req_t` type
 * of the corresponding IP.
 * ```
 */
typedef struct dt_pwrmgr_reset_req_src {
  dt_instance_id_t inst_id; /**< Instance ID of the source of this reset request. */
  size_t reset_req; /**< Index of the reset request signal for that instance. */
} dt_pwrmgr_reset_req_src_t;


/**
 * Get the number of peripheral reset requests.
 *
 * @param dt Instance of pwrmgr.
 * @return Number of reset requests.
 */
size_t dt_pwrmgr_reset_request_count(dt_pwrmgr_t dt);

/**
 * Get the description of a reset request.
 *
 * The reset requests are ordered as they appear in the registers.
 *
 * @param dt Instance of pwrmgr.
 * @param idx Index of the reset request source, between 0 and `dt_pwrmgr_reset_request_count(dt)-1`.
 * @return Description of the reset.
 */
dt_pwrmgr_reset_req_src_t dt_pwrmgr_reset_request(dt_pwrmgr_t dt, size_t idx);
```

On the implementation side, new fields are added to dt to encode this information. This is the table for earlgrey:
```c
// In dt_pwrmgr.c (earlgrey):
static const dt_desc_pwrmgr_t pwrmgr_desc[kDtPwrmgrCount] = {
  [kDtPwrmgrAon] = {
    .inst_id = kDtInstanceIdPwrmgrAon,
    .base_addr = {
      [kDtPwrmgrRegBlockCore] = 0x40400000,
    },
    .first_irq = kTopEarlgreyPlicIrqIdPwrmgrAonWakeup,
    .first_alert = kTopEarlgreyAlertIdPwrmgrAonFatalFault,
    .clock = {
      [kDtPwrmgrClockClk] = kDtClockIoDiv4,
      [kDtPwrmgrClockSlow] = kDtClockAon,
      [kDtPwrmgrClockLc] = kDtClockIoDiv4,
      [kDtPwrmgrClockEsc] = kDtClockIoDiv4,
    },
    .ext = {
      .wakeup_src = {
        [0] = {
          .inst_id = kDtInstanceIdSysrstCtrlAon,
          .wakeup = kDtSysrstCtrlWakeupWkupReq,
        },
        [1] = {
          .inst_id = kDtInstanceIdAdcCtrlAon,
          .wakeup = kDtAdcCtrlWakeupWkupReq,
        },
        [2] = {
          .inst_id = kDtInstanceIdPinmuxAon,
          .wakeup = kDtPinmuxWakeupPinWkupReq,
        },
        [3] = {
          .inst_id = kDtInstanceIdPinmuxAon,
          .wakeup = kDtPinmuxWakeupUsbWkupReq,
        },
        [4] = {
          .inst_id = kDtInstanceIdAonTimerAon,
          .wakeup = kDtAonTimerWakeupWkupReq,
        },
        [5] = {
          .inst_id = kDtInstanceIdSensorCtrlAon,
          .wakeup = kDtSensorCtrlWakeupWkupReq,
        },
      },
      .rst_reqs = {
        [0] = {
          .inst_id = kDtInstanceIdSysrstCtrlAon,
          .reset_req = kDtSysrstCtrlResetReqRstReq,
        },
        [1] = {
          .inst_id = kDtInstanceIdAonTimerAon,
          .reset_req = kDtAonTimerResetReqAonTimer,
        },
      },
    },
  },
};
```

machshev commented on util/dttool.py:
```diff
@@ -7,17 +7,68 @@
+        ext_cls = obj
+
+    if ext_cls is None:
+        logging.error("extension does not define any extension classs")
```
typo
```suggestion
        logging.error("extension does not define any extension class")
```

machshev commented on util/dtgen/helper.py:
```diff
@@ -456,14 +501,17 @@ class IpHelper:
-    def __init__(self, top_helper: TopHelper, ip: IpBlock, default_node: str,
-                 enum_type, array_mapping_type):
+    def __init__(self, top_helper: TopHelper, ip: IpBlock, ipconfig, default_node: str,
+                 enum_type, array_mapping_type, extension: Extension = None):
```
Typing issue: `extension` is typed as non optional `Extension` and assigned `None`. It should be either `Optional[Extension]`, `Union[Extension]`, or some default `Extension` object.

machshev commented on util/dtgen/helper.py:
```diff
@@ -196,6 +197,50 @@ def render_value(self, value: dict[Name, object]) -> str:
+        fields for a given IP. This method MUST not modify `ip_helper` but
+        it can access its public fields.
+        """
+        pass
```
```suggestion
```
`pass` is not required when a docstring is present.

machshev commented on util/dtgen/helper.py:
```diff
@@ -196,6 +197,50 @@ def render_value(self, value: dict[Name, object]) -> str:
+        Override this function to return the name of the IP to consider.
+        """
+
+    def extend_dt_ip(self, ip_helper: "IpHelper") -> "StructType":
```
Quoted type annotation is only required if the type isn't defined at the point when the line is executed i.e. method that returns an object of the class the method belongs to.

```suggestion
    def extend_dt_ip(self, ip_helper: "IpHelper") -> StructType:
```

machshev commented on util/dtgen/helper.py:
```diff
@@ -196,6 +197,50 @@ def render_value(self, value: dict[Name, object]) -> str:
+        Override this function to return the name of the IP to consider.
+        """
+
+    def extend_dt_ip(self, ip_helper: "IpHelper") -> "StructType":
```
Method is non abstract, which means it will be inherited by sub classes without having to overload, and can be called. Which means the typing is incorrect as this default implementation returns `None` (there is always an object returned from every function called, without explicit return the returned object is `None`).

So for the typing to be correct it needs to either:
- return a default StructType object of some sort.
- update return type to allow `None` i.e. `Union[StructType]`
- or make the method abstract so it has to be implemented in sub classes

pamaury replied:
I made it non-abstract on purpose. My reasoning is that extensions may gain features in the future, ie more hooks. We don't want that adding a new hooks break existing plugins. Therefore I only made one method abstract which is the only one where no default makes sense.

matutem commented on hw/ip_templates/pwrmgr/util/ipconfig.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        self.param_values = ipconfig["param_values"]
+
+    def wakeup_list(self) -> List[Dict]:
```
For ipgen we will want iterables rather than collections.

matutem replied:
Sorry for the delay. For ipgen we tpically iterate, so returning iterators would be marginally better.

pamaury commented on hw/ip_templates/pwrmgr/util/ipconfig.py:
```diff
@@ -0,0 +1,53 @@
+
+
+class IpConfig:
+    def __init__(self, ipconfig: object):
```
Ignoring the bazel things, yes in `dt.py` (which is the sole user), I set the `sys.path` to include just this directory (it's kind of like a local import). Also an annoying detail is that with ipgen this file is copied in each top's directory...
I am unfortunately not very good at python so there might be a better of doing this.

matutem replied:
I was expecting sys.path manipulations. It would make sense to block copying ip_templates/*/util in ipgen.

a-will commented on util/dtgen/helper.py:
```diff
@@ -547,6 +616,28 @@ def _init_clocks(self):
+    def has_reset_requests(self):
+        return len(self.reset_req_map) > 0
+
+    def _init_reset_requests(self):
+        self.reset_req_enum = self._enum_type(Name([]), Name(["dt"]) + self.ip_name +
+                                              Name(["reset", "req"]))
+        self.reset_req_map = OrderedDict()
```
Random aside: a "reset request" network seems like an atypical thing that only OT does, hehe.

matutem replied:
Maybe because resets are more carefully handled? There are several reset requests in other systems, POR, magic reset key presses, watchdogs...

a-will commented on rules/opentitan/hw.bzl:
```diff
@@ -71,7 +128,13 @@ describe_top_rule = rule(
-        "ip_hjson": attr.string_keyed_label_dict(allow_files = True, doc = "mapping from hjson files to IP name"),
+        "ip_map": attr.string_keyed_label_dict(
+            allow_files = True,
+            # Transition to host becomes some of those attributes could targets (e.g. python libraries) that only work
+            # on the host platform.
```
nit: Seems like a few typos were here.

```suggestion
            # Transition to host because some of those attributes could use targets (e.g. python libraries) that only work
            # on the host platform.
```

a-will reviewed (approved):
This seems fine as a first attempt, as long as we're not exposing this as public API. It gives a little too much power to the user with an awkward file placement + template interface, which makes it rather difficult to implement changes to the format as needed.

At some point, I suspect we'll figure out a property API + an input to the generator that is more declarative in nature.

matutem commented on hw/ip_templates/pwrmgr/util/dt.py:
```diff
@@ -0,0 +1,184 @@
+import sys
+
+sys.path.insert(0, os.path.dirname(os.path.realpath(__file__)))
+from ipconfig import IpConfig  # noqa: E402
```
Please don't use the name IpConfig: there is a class in util/ipgen/lib.py by that name.

pamaury replied:
renamed to `PwrmgrIpConfig`

matutem commented on hw/ip_templates/pwrmgr/util/dt.py:
```diff
@@ -0,0 +1,184 @@
+"""This contains a class which is used to help generate the device tables (DT)
+files.
+"""
+from dtgen.helper import IpHelper, Extension, StructType, ScalarType, ArrayMapType
```
Please run the linter on this python code: it should complain about import ordering, and who knows what else.

pamaury replied:
The linter is already running on it but we are currently switching to a new linter with stricter checks.

matutem commented on hw/ip_templates/pwrmgr/util/dt.py:
```diff
@@ -0,0 +1,184 @@
+            raise RuntimeError("the pwrmgr extension requires the ipconfig to be provided")
+        self.ipconfig = IpConfig(self.ip_helper.ipconfig)
+
+    def create_ext(ip_helper: IpHelper):
```
Nit: perhaps annotate with @classmethod or @staticmethod?

matutem commented on hw/ip_templates/pwrmgr/util/dt.py:
```diff
@@ -0,0 +1,184 @@
+        )
+        return st
+
+    def fill_dt_ip(self, m) -> Optional[dict]:
```
"m" type? I don't see it used, and could you give it a more descriptive name?

pamaury replied:
It's documented in the `Extension` class in `helper.py` but ok

matutem commented on hw/ip_templates/pwrmgr/util/ipconfig.py:
```diff
@@ -0,0 +1,53 @@
+class IpConfig:
+    def __init__(self, ipconfig: object):
+        """
+        Initialize an `IpConfig` from an already loaded and parsed `ipconfog.hjson`
```
```suggestion
        Initialize an `IpConfig` from an already loaded and parsed `ipconfig.hjson`
```

matutem commented on hw/ip_templates/pwrmgr/util/ipconfig.py:
```diff
@@ -0,0 +1,53 @@
+        - `name`: name of the wakeup .
+        - `module`: the module where the wakeup comes from.
+
+        The list is ordered in the same as in the description.
```
```suggestion
        The list is ordered as in the description.
```

matutem commented on hw/ip_templates/pwrmgr/util/ipconfig.py:
```diff
@@ -0,0 +1,53 @@
+        - `name`: name of the reset request.
+        - `module`: the module where the reset comes from.
+
+        The list is ordered in the same as in the description.
```
Like above

matutem commented on hw/top/BUILD:
```diff
@@ -127,6 +127,30 @@ opentitan_top_dt_api(
+    hdrs = flatten([
+        opentitan_if_ip(
+            ip,
+            ["dt_{}_hdr".format(ip)],
```
Why not f-strings? Way more ergonomic in my opinion.

pamaury replied:
Not available in bazel :(

matutem commented on pyproject.toml:
```diff
@@ -190,6 +190,7 @@ extend-exclude = [
   "util/vendor.py",
   "util/verible-format.py",
   "util/version_file.py",
+  "**/pwrmgr/util/*.py",
```
hw/ip_templates/*/util/*.py ?

matutem commented on rules/opentitan/hw.bzl:
```diff
@@ -4,18 +4,31 @@
+        name = "pwrmgr",
+        files = {
+            'hjson': "//hw/top_earlgrey/ip_autogen/pwrmgr:data/pwrmgr.hjson",
+            'ipconfig': "//hw/top_earlgrey/ip_autogen/pwrmgr:data/top_earlgrey_pwrmgr.ipconfig.hjson",
```
Nit: naming the ipconfig.hjson files starting with the top seems silly, or as the "go" folks would say, it is a stutter. It is obviously not this PR's business.

pamaury replied:
Agreed...

matutem commented on rules/opentitan/hw.bzl:
```diff
@@ -43,26 +56,70 @@ OpenTitanTopInfo = provider(
+    - "target": return the raw bazel target
+    """
+    if ipname not in top.ip_map:
+        fail("top {} does not contain IP {}".format(top.name, ipname))
```
Nit: f-strings

matutem commented on util/dtgen/dt_ip.c.tpl:
```diff
@@ -20,10 +24,23 @@ ${helper.inst_struct.render_type_def()}
 <%
   dt_array_name = helper.ip_name + Name(["desc"])
   dt_array = dt_array_name.as_snake_case()
+  dt_array_count = helper.inst_enum.name + Name(["count"])
```
Curious: what does Name(["count"]) expand to? Why not just "count".

pamaury replied:
This is just a technical detail: for consistency I use `Name` everywhere and only render them at the point of use. So yes this could be replaced by `dt_array_count = helper.inst_enum.name.as_c_enum() + "Count"` but I prefer not to hardcode that if possible.

matutem commented on util/dtgen/dt_ip.c.tpl:
```diff
@@ -20,10 +24,23 @@ ${helper.inst_struct.render_type_def()}
+ * `dt` if it's a valid index. Otherwise, this macro will `return` (i.e. exit
+ * the function) with the provided default value.
+ */
+#define TRY_GET_DT(dt, default) \
```
This is C code, but is there a better way to handle a bad situation?

a-will replied:
No, this is the correct way to handle it. It should return whatever is the error / unknown that is specified to the macro.

pamaury replied:
This is a trade-off: I don't want the DT function to be like DIFs with an error code, this would make the code too heavy and potentially costly in some places. Under normal usage, it should be impossible to call a DT function with bad parameters. Thus the decision to simply return a valid but documented value when the arguments are invalid.

matutem commented on util/dtgen/helper.py:
```diff
@@ -141,6 +146,7 @@ def field_type(self, name: Name) -> object:
         return self.fields[name][0]
 
     def as_c_type(self) -> str:
+        assert self.name is not None, "can get the name of an anonymous StructType"
```
```suggestion
        assert self.name is not None, "cannot get the name of an anonymous StructType"
```

matutem commented on util/dtgen/helper.py:
```diff
@@ -456,12 +519,14 @@ class IpHelper:
 
-    def __init__(self, top_helper: TopHelper, ip: IpBlock, default_node: str,
-                 enum_type, array_mapping_type):
+    def __init__(self, top_helper: TopHelper, ip: IpBlock, ipconfig, default_node: str,
```
Missing types.

pamaury replied:
I have added them but it's quite unhelpful because they can be anything

matutem commented on util/dtgen/helper.py:
```diff
@@ -447,6 +501,15 @@ def _init_dev_type_map(self):
                 self.dev_type_values[Name.from_snake_case(m["name"])] = \
                     Name.from_snake_case(module_name)
 
+    def get_module_type(self, module_name) -> str:
```
module_name type?

matutem commented on util/dtgen/helper.py:
```diff
@@ -178,24 +184,72 @@ def render_var_decl(self, name: Name) -> str:
         """
         Render a value which is a dictionary mapping fields to value.
         """
+        assert isinstance(value, Mapping), "StructType can only render values which are mappings"
```
Why not set Mapping as the type in the function decl above?

matutem commented on pyproject.toml:
```diff
@@ -190,6 +190,7 @@ extend-exclude = [
   "util/vendor.py",
   "util/verible-format.py",
   "util/version_file.py",
+  "**/pwrmgr/util/*.py",
```
```suggestion
  "hw/ip_templates/pwrmgr/util/*.py",
```
To avoid dealing with replicates.

pamaury replied:
It is important to exclude the replicates because they are copies and therefore will fail the linting as well.

pamaury commented on util/dtgen/helper.py:
```diff
@@ -178,24 +184,72 @@ def render_var_decl(self, name: Name) -> str:
+        return `None`.
+        """
+
+    def extend_dt_ip(self) -> Optional[StructType]:
```
I didn't make them abstract because it's possible that an extension wants to add code to the template without extending the DT structure. This way a plugin can extend only what it wants and not redefine empty functions for no reason.

matutem replied:
Good point.

pamaury commented on util/dtgen/dt_ip.c.tpl:
```diff
@@ -4,14 +4,18 @@
 // Device table API auto-generated by `dtgen`
 <%
-from topgen.lib import Name, is_top_reggen, is_ipgen
+from topgen.lib import Name
```
Probably yes

pamaury commented:
I have addressed the comments and also changed the API a bit. It still exposes the same information but now using structures which should make it clearer and harder to misused.

matutem commented:
Too bad this keeps an abstraction named ipconfig since that already exists, perhaps dtconfig or just config is a better name? Also, it seems the pwrmgr python code is top-independent, and the top dependency is only introduced by the ipgen-generated pwrmgr.ipconfig.hjson file. This argues for blocking the copy of this code by ipgen. While that can be done in a separate PR, I would expect ci/lint/ruff.toml would just look at hw/ip_templates/pwrmgr/... You agreed in https://github.com/lowRISC/opentitan/pull/26248#discussion_r1962655957, but I am confused since that comment seems to refer to pyproject.toml: is that a pseudonym for ruff.toml?

matutem reviewed (approved):
LGTM with some pending issues, most important the ipconfig name.

pamaury commented:
> Too bad this keeps an abstraction named ipconfig since that already exists, perhaps dtconfig or just config is a better name? Also, it seems the pwrmgr python code is top-independent, and the top dependency is only introduced by the ipgen-generated pwrmgr.ipconfig.hjson file. This argues for blocking the copy of this code by ipgen. While that can be done in a separate PR, I would expect ci/lint/ruff.toml would just look at hw/ip_templates/pwrmgr/... You agreed in [#26248 (comment)](https://github.com/lowRISC/opentitan/pull/26248#discussion_r1962655957), but I am confused since that comment seems to refer to pyproject.toml: is that a pseudonym for ruff.toml?

Hi Matute, I think you misunderstood the change to `ruff.toml`: this is an exclude list, not a lint list. The new lint is a *lot* more stringent that the old one which is why I added those files to the exclude list. Otherwise the linter will complain that the various copies (in top_earlgrey, etc) fail linting. I agree that avoiding the copy in the first place would be a better solution but I am not sure what the best way to achieve this. Let's discuss this offline.

pamaury commented:
I am merging this. For completeness and to summarize extensive discussions I had notably with @matutem and @a-will : 
- The behaviour of copying the library and dt files for each top might or might not be good. In any case ipgen always copies everything and the point could be made that copying everything is kind of the point so that the generated RTL does not depend on the template. While this might hold for hardware artefacts, things are less clear for software. This could become a moot point if the second issue below is solved.
- The API for the plugins is way too open: we are currently allowing an arbitrary python script that can access the entire repo, and in particular all of `helper.py`. We definitely do not want to make this API public because it will be impossible to  maintain backwards compatibility. My plan is to develop a more declarative way of creating extension based on a Hjson description + a Mako template. But before doing that, it is important to see the features that extensions need.

matutem commented:
A minor clarification: currently ipgen doesn't copy everything. It doesn't copy the .tpldesc.hjson file, and it would be perfectly understandable (and easy to support) if we decide to avoid copying these support files.
